sklearn/utils/tests/test_validation.py

```diff
@@ -218,6 +219,25 @@ def test_check_array():
     assert_true(isinstance(result, np.ndarray))
 
 
+def test_check_array_pandas_dtype_object_conversion():
+    # test that data-frame like objects with dtype object
+    # get converted
+    X = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=np.object)
+    X_df = MockDataFrame(X)
+    assert_equal(check_array(X_df).dtype.kind, "f")
+    assert_equal(check_array(X_df, ensure_2d=False).dtype.kind, "f")
+    # smoke-test against dataframes with column named "dtype"
+    X_df.dtype = "Hans"
+    assert_equal(check_array(X_df, ensure_2d=False).dtype.kind, "f")
```

Comment:
I am a bit worried that the above is accepting a very weird object that should not be valid: the "dtype" column is not an iterable with the right length. I would expect check_array to fail.